tstest/integration/vms/vms_test.go

```diff
@@ -424,140 +428,126 @@ func TestVMIntegrationEndToEnd(t *testing.T) {
+				}
+
 				t.Parallel()
 
+				err := ramsem.Acquire(ctx, int64(distro.mem))
+				if err != nil {
+					t.Fatalf("can't acquire ram semaphore: %v", err)
+				}
+				defer ramsem.Release(int64(distro.mem))
+
 				cancel := mkVM(t, n, distro, string(pubkey), loginServer, dir)
-				cancels <- cancel
+				defer cancel()
+				var ipm ipMapping
+
+				t.Run("wait-for-start", func(t *testing.T) {
+					waiter := time.NewTicker(time.Second)
+					defer waiter.Stop()
+					var ok bool
+					for {
+						<-waiter.C
+						ipMu.Lock()
+						if ipm, ok = ipMap[distro.name]; ok {
+							ipMu.Unlock()
+							break
+						}
+						ipMu.Unlock()
+					}
+				})
+
+				testDistro(t, loginServer, signer, ipm)
 			})
 		}
 	})
+}
 
-	close(cancels)
-	for cancel := range cancels {
-		//lint:ignore SA9001 They do actually get ran
-		defer cancel()
-
-		if len(cancels) == 0 {
-			t.Log("all VMs started")
+func testDistro(t *testing.T, loginServer string, signer ssh.Signer, ipm ipMapping) {
+	t.Helper()
+	port := ipm.port
+	hostport := fmt.Sprintf("127.0.0.1:%d", port)
+	ccfg := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer), ssh.Password(securePassword)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	// NOTE(Xe): This retry loop helps to make things a bit faster, centos sometimes is slow at starting its sshd. I don't know why they don't use socket activation.
+	const maxRetries = 5
+	var working bool
+	for i := 0; i < maxRetries; i++ {
+		cli, err := ssh.Dial("tcp", hostport, ccfg)
+		if err == nil {
+			working = true
+			cli.Close()
 			break
 		}
-	}
 
-	t.Run("wait-for-vms", func(t *testing.T) {
-		t.Log("waiting for VMs to register")
-		waiter := time.NewTicker(time.Second)
-		defer waiter.Stop()
-		n := 0
-		for {
-			<-waiter.C
-			ipMu.Lock()
-			if len(ipMap) == numDistros {
-				ipMu.Unlock()
-				break
-			} else {
-				if n%30 == 0 {
-					t.Logf("ipMap:   %d", len(ipMap))
-					t.Logf("distros: %d", numDistros)
-				}
-			}
-			n++
-			ipMu.Unlock()
-		}
-	})
-
-	ipMu.Lock()
-	defer ipMu.Unlock()
-	t.Run("join-net", func(t *testing.T) {
-		for _, ipm := range ipMap {
-			ipm := ipm
-			port := ipm.port
-			t.Run(ipm.name, func(t *testing.T) {
-				tstest.FixLogs(t)
-				t.Parallel()
-
-				hostport := fmt.Sprintf("127.0.0.1:%d", port)
-
-				// NOTE(Xe): This retry loop helps to make things a bit faster, centos sometimes is slow at starting its sshd. I don't know why they don't use socket activation.
-				const maxRetries = 5
-				var working bool
-				for i := 0; i < maxRetries; i++ {
-					conn, err := net.Dial("tcp", hostport)
-					if err == nil {
-						working = true
-						conn.Close()
-						break
-					}
-
-					time.Sleep(5 * time.Second)
-				}
-
-				if !working {
-					t.Fatalf("can't connect to %s, tried %d times", hostport, maxRetries)
-				}
+		time.Sleep(10 * time.Second)
```

Comment:
10s is maybe a bit slow here. Make it 500ms and use a deadline instead of a fixed retry count?

Reply:
This is intended to work around CentOS brokenness :( CentOS will slay all SSH sessions on transition to multi-user.target. I can make it into a deadline loop in a later patch.